### PR TITLE
Fix corner case in pearson correlation

### LINKED ISSUE #2920: Incorrect Pearson Correlation Output for Constant y_true in torchmetrics
## 🐛 Bug

When computing the Pearson correlation using `torchmetrics.PearsonCorrCoef`, it returns `1` for some cases where `y_true` is constant. Mathematically, the Pearson correlation should be undefined (or return NaN), since the standard deviation of `y_true` is zero. In contrast, `scipy.stats.pearsonr` returns`0` in such cases.

### To Reproduce

```python
import torchmetrics
import torch
from scipy import stats

y_pred = torch.tensor([[-0.1816,  0.6568,  0.9788, -0.1425], [-0.4111,  0.3940,  1.4834,  0.1322]])
y_true = torch.tensor([[4.0268,  5.9401,  1.0000,  1.0000], [6.4956,  5.6684,  1.0000,  1.0000]])

pearson_corr = torchmetrics.PearsonCorrCoef(num_outputs=4)
print('Torchmetrics: ', pearson_corr(y_pred, y_true))
print('Scipy: ', stats.pearsonr(y_pred, y_true))
```

Returns:
```python
Torchmetrics:  tensor([-1.0000,  1.0000,  1.0000,     nan])
Scipy:  PearsonRResult(statistic=array([-1.,  1.,  0.,  0.], dtype=float32), pvalue=array([1., 1., 1., 1.], dtype=float32))
```

### Expected behavior

For columns where `y_true` is constant (e.g., y_true[:, 2] = [1, 1]), the Pearson correlation should return `nan`.

### Environment

- TorchMetrics version (if build from source, add commit SHA): 1.6.1 
- Python: 3.11.0
- Torch: 2.5.1

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #2920

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2975.org.readthedocs.build/en/2975/

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: src/torchmetrics/functional/regression/pearson.py [hunk 1]
```diff
@@ -106,12 +106,20 @@ def _pearson_corrcoef_compute(
...
-    return torch.clamp(corrcoef, -1.0, 1.0)
+    zero_var_mask = (var_x < bound) | (var_y < bound)
+    corrcoef = torch.full_like(corr_xy, float("nan"), device=corr_xy.device, dtype=corr_xy.dtype)
+    valid_mask = ~zero_var_mask
+
+    if valid_mask.any():
+        corrcoef[valid_mask] = (
+            (corr_xy[valid_mask] / (var_x[valid_mask] * var_y[valid_mask]).sqrt()).squeeze().to(corrcoef.dtype)
+        )
+        corrcoef = torch.clamp(corrcoef, -1.0, 1.0)
```
[the remark this answers is not in the document]
Comment (justusschock) [reply src/torchmetrics/functional/regression/pearson.py:121]: let's not do this @Borda , we already squeeze above with masking , so squeezing here is not necessary at all (since clamping preserves the already squeezed shape)

**Review by justusschock — APPROVED** [review]

File: src/torchmetrics/functional/regression/pearson.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply src/torchmetrics/functional/regression/pearson.py:121]: yeah, but then squeezing has to be done below, not here as this is a special  case only.
